### PR TITLE
GUVNOR-2827: [Guided Decision Table] Not possible to create two conditions based on the same negated pattern

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/GUVNOR-2827

### CODE REVIEW

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java [hunk 1]
```diff
@@ -344,540 +332,572 @@ private void addAction( IAction action,
...
+                isValid = Boolean.TRUE.equals(dcv.getBooleanValue());
             }
 
-            if ( isValid ) {
+            if (isValid) {
 
                 // get or create the pattern it belongs too
-                IPattern ifp = findByFactPattern( patterns,
-                                                  pattern.getBoundName() );
+                IPattern ifp = findByFactPattern(patterns,
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java:674]: Here's the change!

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java [hunk 2]
```diff
@@ -344,540 +332,572 @@ private void addAction( IAction action,
...
-    private boolean no( String operator ) {
-        return operator == null || "".equals( operator );
+    private boolean no(String operator) {
+        return operator == null || "".equals(operator);
     }
 
-    private IPattern findByFactPattern( List<IPattern> patterns,
-                                        String boundName ) {
-        if ( boundName == null || boundName.isEmpty() ) {
+    private IPattern findByFactPattern(List<IPattern> patterns,
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java:770]: ..and here.

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java [hunk 3]
```diff
@@ -344,540 +332,572 @@ private void addAction( IAction action,
...
             }
         }
         return null;
     }
 
-    void doAttribs( List<BaseColumn> allColumns,
-                    List<AttributeCol52> attributeCols,
-                    List<DTCellValue52> row,
-                    RuleModel rm ) {
+    private IPattern doFindByNegatedFactPattern(final List<IPattern> patterns,
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java:807]: ..and here.

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/test/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistenceTest.java [hunk 4]
```diff
@@ -960,621 +956,675 @@ public void multipleLHSNotPatternInclusion() {
...
                 "end\n";
 
-        assertEqualsIgnoreWhitespace( expectedDecisionTableDrl,
-                                      actualDecisionTableDrl );
+        assertEqualsIgnoreWhitespace(expectedDecisionTableDrl,
+                                     actualDecisionTableDrl);
+    }
+
+    @Test
+    public void multipleConstraintLHSNotPatternInclusion() {
```
Comment (manstis) [inline drools-workbench-models/drools-workbench-models-guided-dtable/src/test/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistenceTest.java:964]: New test.

**Comment by manstis** [comment]
@Rikkola @karreiro I've commented where the actual changes are.. otherwise it's all formatting.

**Comment by manstis** [comment]
Part of an ensemble:-

- https://github.com/droolsjbpm/drools/pull/1105
- https://github.com/droolsjbpm/drools-wb/pull/399

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java [hunk 5]
```diff
@@ -344,540 +332,572 @@ private void addAction( IAction action,
...
+                FactPattern fp = (FactPattern) ifp;
+                if (fp.isNegated()) {
+                    if (fp.getFactType().equals(factType)) {
+                        return fp;
+                    }
+                }
+            } else if (ifp instanceof FromEntryPointFactPattern) {
+                FromEntryPointFactPattern fefp = (FromEntryPointFactPattern) ifp;
+                FactPattern fp = fefp.getFactPattern();
+                if (fp.isNegated()) {
```
Comment (jomarko) [inline drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java]: @manstis I think here is nice opportunity to use single method for the blocks starting at lines 812 and 820. WDYT?

Comment (manstis) [reply drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java]: @jomarko Updated... although arguable less readable now ;-)

Comment (jomarko) [reply drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java]: @manstis sure, I agree that my proposed change didn't bring readability enhancement, however I think this was the case, when updates in one block would implicate the same update in the second block so from my point of view is maintaining this piece of code at single place important.

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java [hunk 6]
```diff
@@ -344,540 +333,582 @@ private void addAction( IAction action,
...
+                }
+            }
+        }
+        return null;
+    }
+
+    private Optional<IPattern> checkNegatedFactPattern(final FactPattern fp,
+                                                       final String factType) {
+        if (fp.isNegated()) {
+            if (fp.getFactType().equals(factType)) {
```
Comment (karreiro) [inline drools-workbench-models/drools-workbench-models-guided-dtable/src/main/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistence.java:834]: In my personal opinion, keeping only one `if` (with `&&`) would be more expressive.

File: drools-workbench-models/drools-workbench-models-guided-dtable/src/test/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistenceTest.java (hunk 4 above)
Comment (karreiro) [reply drools-workbench-models/drools-workbench-models-guided-dtable/src/test/java/org/drools/workbench/models/guided/dtable/backend/GuidedDTDRLPersistenceTest.java:964]: I would extract these blocks to methods (I think that we could avoid some duplication). However it's pretty readable this way too (it's just a personal opinion).

**Review by karreiro — APPROVED** [review]
Nice to see the use of the `Optional` becoming more frequent :-)

Thanks, Mike!

**Pull request merged**